File: apps/api/src/discover/discover.service.ts
import { Injectable } from '@nestjs/common';
import { inArray } from 'drizzle-orm';
import { DatabaseService } from '@/db/database.service';
import {
  games,
  domainEvents,
  DiscoverCandidate,
  DiscoverApplyResult,
} from '@workspace/api-contract';
import { IgdbService } from '@/games/igdb.service';
import { GamesService } from '@/games/games.service';

@Injectable()
export class DiscoverService {
  constructor(
    private readonly databaseService: DatabaseService,
    private readonly igdbService: IgdbService,
    private readonly gamesService: GamesService,
  ) {}

  async scan(
    count: number,
    actorId: string,
  ): Promise<{
    scanEventId: number;
    candidates: DiscoverCandidate[];
    totalReturned: number;
    alreadyAddedCount: number;
  }> {
    const igdbResults = await this.igdbService.discoverCandidates(count);

    const igdbIds = igdbResults.map((g) => g.id);

    let existingSet = new Set<number>();
    if (igdbIds.length > 0) {
      const existingRows = await this.databaseService.db
        .select({ igdbId: games.igdbId })
        .from(games)
        .where(inArray(games.igdbId, igdbIds));
      existingSet = new Set(existingRows.map((r) => r.igdbId));
    }

    const candidates: DiscoverCandidate[] = igdbResults.map((g) => {
      const rawUrl = g.cover?.url;
      const imageId = g.cover?.image_id;
      let coverUrl: string | null = null;
      if (imageId) {
        coverUrl = `https://images.igdb.com/igdb/image/upload/t_cover_big/${imageId}.jpg`;
      } else if (rawUrl) {
        const fullUrl = rawUrl.startsWith('//') ? `https:${rawUrl}` : rawUrl;
        coverUrl = fullUrl.replace('t_thumb', 't_cover_big');
      }

      return {
        igdbId: g.id,
        name: g.name,
        firstReleaseDate: g.first_release_date ?? null,
        coverUrl,
        totalRating: g.total_rating ?? null,
        totalRatingCount: g.total_rating_count ?? null,
        genres:
          g.genres?.map((genre) => genre.name ?? '').filter(Boolean) ?? [],
        platforms: g.platforms?.map((p) => p.name ?? '').filter(Boolean) ?? [],
        isAlreadyAdded: existingSet.has(g.id),
      };
    });

    const totalReturned = candidates.length;
    const alreadyAddedCount = candidates.filter((c) => c.isAlreadyAdded).length;

    const eventPayload = {
      count,
      candidates: candidates.map((c) => ({
        igdbId: c.igdbId,
        name: c.name,
        isAlreadyAdded: c.isAlreadyAdded,
      })),
      totalReturned,
      alreadyAddedCount,
    };

    const [insertedEvent] = await this.databaseService.db
      .insert(domainEvents)
      .values({
        eventType: 'discover_games.scanned',
        actorId,
        payload: eventPayload,
      })
      .returning({ id: domainEvents.id });

    return {
      scanEventId: insertedEvent.id,
      candidates,
      totalReturned,
      alreadyAddedCount,
    };
  }

  async apply(
    selectedIgdbIds: number[],
    scanEventId: number,
    actorId: string,
  ): Promise<{
    success: boolean;
    applyEventId: number;
    results: DiscoverApplyResult[];
  }> {
    const results: DiscoverApplyResult[] = [];

    for (const igdbId of selectedIgdbIds) {
      try {
        const syncResult = await this.gamesService.syncGameByIgdbId(igdbId);
        if (syncResult) {
          results.push({
            igdbId,
            name: syncResult.game.name,
            status: syncResult.operation,
            error: null,
          });
        } else {
          results.push({
            igdbId,
            name: null,
            status: 'error',
            error: 'Game not found on IGDB',
          });
        }
      } catch (err) {
        results.push({
          igdbId,
          name: null,
          status: 'error',
          error: err instanceof Error ? err.message : 'Unknown error',
        });
      }
    }

    const eventPayload = {
      scanEventId,
      selectedIgdbIds,
      results,
    };

    const [insertedEvent] = await this.databaseService.db
      .insert(domainEvents)
      .values({
        eventType: 'discover_games.applied',
        actorId,
        payload: eventPayload,
      })
      .returning({ id: domainEvents.id });

    await this.gamesService.refreshAllGamesView();

    return {
      success: true,
      applyEventId: insertedEvent.id,
      results,
    };
  }
}

File: apps/api/services/discover_service.go
package services

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"gaeldle/api/models"
)

type DiscoverService struct {
	db           *sql.DB
	igdbService  *IgdbService
	gamesService *GamesService
}

func NewDiscoverService(db *sql.DB, igdbService *IgdbService, gamesService *GamesService) *DiscoverService {
	return &DiscoverService{
		db:           db,
		igdbService:  igdbService,
		gamesService: gamesService,
	}
}

type DiscoverScanResult struct {
	ScanEventID       int                        `json:"scanEventId"`
	Candidates        []models.DiscoverCandidate `json:"candidates"`
	TotalReturned     int                        `json:"totalReturned"`
	AlreadyAddedCount int                        `json:"alreadyAddedCount"`
}

func (s *DiscoverService) Scan(count int, actorID string) (*DiscoverScanResult, error) {
	if count <= 0 {
		count = 10
	}
	if count > 50 {
		count = 50
	}

	igdbResults, err := s.igdbService.DiscoverCandidates(count)
	if err != nil {
		return nil, fmt.Errorf("failed to discover candidates from IGDB: %w", err)
	}

	var igdbIDs []int
	for _, g := range igdbResults {
		igdbIDs = append(igdbIDs, g.ID)
	}

	existingSet := make(map[int]bool)
	if len(igdbIDs) > 0 && s.db != nil {
		var idStrs []string
		for _, id := range igdbIDs {
			idStrs = append(idStrs, strconv.Itoa(id))
		}
		query := fmt.Sprintf("SELECT igdb_id FROM game WHERE igdb_id IN (%s)", strings.Join(idStrs, ","))
		rows, err := s.db.Query(query)
		if err == nil {
			defer rows.Close()
			for rows.Next() {
				var id int
				if err := rows.Scan(&id); err == nil {
					existingSet[id] = true
				}
			}
		}
	}

	var candidates []models.DiscoverCandidate
	alreadyAddedCount := 0

	for _, g := range igdbResults {
		var coverURL *string
		if g.Cover != nil {
			if g.Cover.ImageID != "" {
				u := fmt.Sprintf("https://images.igdb.com/igdb/image/upload/t_cover_big/%s.jpg", g.Cover.ImageID)
				coverURL = &u
			} else if g.Cover.URL != "" {
				full := g.Cover.URL
				if strings.HasPrefix(full, "//") {
					full = "https:" + full
				}
				u := strings.ReplaceAll(full, "t_thumb", "t_cover_big")
				coverURL = &u
			}
		}

		var genres []string
		for _, gen := range g.Genres {
			if gen.Name != "" {
				genres = append(genres, gen.Name)
			}
		}

		var platforms []string
		for _, p := range g.Platforms {
			if p.Name != "" {
				platforms = append(platforms, p.Name)
			}
		}

		isAdded := existingSet[g.ID]
		if isAdded {
			alreadyAddedCount++
		}

		candidates = append(candidates, models.DiscoverCandidate{
			IgdbID:           g.ID,
			Name:             g.Name,
			FirstReleaseDate: g.FirstReleaseDate,
			CoverURL:         coverURL,
			TotalRating:      g.TotalRating,
			TotalRatingCount: g.TotalRatingCount,
			Genres:           genres,
			Platforms:        platforms,
			IsAlreadyAdded:   isAdded,
		})
	}

	scanEventID := 0
	if s.db != nil {
		eventPayload, _ := json.Marshal(map[string]interface{}{
			"count":             count,
			"candidates":        candidates,
			"totalReturned":     len(candidates),
			"alreadyAddedCount": alreadyAddedCount,
		})

		_ = s.db.QueryRow(`
			INSERT INTO domain_events (event_type, actor_id, payload)
			VALUES ($1, $2, $3)
			RETURNING id
		`, "discover_games.scanned", actorID, string(eventPayload)).Scan(&scanEventID)
	}

	return &DiscoverScanResult{
		ScanEventID:       scanEventID,
		Candidates:        candidates,
		TotalReturned:     len(candidates),
		AlreadyAddedCount: alreadyAddedCount,
	}, nil
}

type DiscoverApplyResponse struct {
	Success      bool                         `json:"success"`
	ApplyEventID int                          `json:"applyEventId"`
	Results      []models.DiscoverApplyResult `json:"results"`
}

func (s *DiscoverService) Apply(selectedIgdbIds []int, scanEventID int, actorID string) (*DiscoverApplyResponse, error) {
	var results []models.DiscoverApplyResult

	for _, igdbID := range selectedIgdbIds {
		res, err := s.gamesService.SyncGameByIgdbId(igdbID, false, actorID)
		if err != nil {
			errStr := err.Error()
			results = append(results, models.DiscoverApplyResult{
				IgdbID: igdbID,
				Name:   nil,
				Status: "error",
				Error:  &errStr,
			})
		} else if res != nil && res.Game != nil {
			name := res.Game.Name
			results = append(results, models.DiscoverApplyResult{
				IgdbID: igdbID,
				Name:   &name,
				Status: res.Operation,
				Error:  nil,
			})
		} else {
			errStr := "Game not found on IGDB"
			results = append(results, models.DiscoverApplyResult{
				IgdbID: igdbID,
				Name:   nil,
				Status: "error",
				Error:  &errStr,
			})
		}
	}

	applyEventID := 0
	if s.db != nil {
		eventPayload, _ := json.Marshal(map[string]interface{}{
			"scanEventId":     scanEventID,
			"selectedIgdbIds": selectedIgdbIds,
			"results":         results,
		})

		_ = s.db.QueryRow(`
			INSERT INTO domain_events (event_type, actor_id, payload)
			VALUES ($1, $2, $3)
			RETURNING id
		`, "discover_games.applied", actorID, string(eventPayload)).Scan(&applyEventID)
	}

	go s.gamesService.RefreshAllGamesView(true)

	return &DiscoverApplyResponse{
		Success:      true,
		ApplyEventID: applyEventID,
		Results:      results,
	}, nil
}
